clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/loadbalancer/UpsertOpenstackLoadBalancerAtomicOperation.groovy

```diff
@@ -0,0 +1,105 @@
+ */
+
+package com.netflix.spinnaker.clouddriver.openstack.deploy.ops.loadbalancer
+
+import com.netflix.spinnaker.clouddriver.openstack.client.OpenstackClientProvider
+import com.netflix.spinnaker.clouddriver.openstack.deploy.description.loadbalancer.OpenstackLoadBalancerDescription
+import com.netflix.spinnaker.clouddriver.openstack.deploy.exception.OpenstackOperationException
+import com.netflix.spinnaker.clouddriver.openstack.domain.LoadBalancerPool
+import com.netflix.spinnaker.clouddriver.openstack.domain.Region
+import com.netflix.spinnaker.clouddriver.openstack.domain.VirtualIP
+import com.netflix.spinnaker.clouddriver.openstack.task.TaskAware
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperations
+import org.openstack4j.model.network.NetFloatingIP
+import org.openstack4j.model.network.ext.LbPool
+import org.openstack4j.model.network.ext.Vip
+
+class UpsertOpenstackLoadBalancerAtomicOperation implements AtomicOperation<Map>, TaskAware {
+  OpenstackLoadBalancerDescription description
+
+  UpsertOpenstackLoadBalancerAtomicOperation(OpenstackLoadBalancerDescription description) {
+    this.description = description
+  }
+
+  /*
+   * curl -X POST -H "Content-Type: application/json" -d  '[ {  "upsertLoadBalancer": { "region": "region", "account": "test", "name": "test",  "protocol": "HTTP", "method" : "ROUND_ROBIN", "subnetId": "9e0d71a9-0086-494a-91d8-abad0912ba83", "externalPort": 80, "internalPort": 8100, "networkId": "9e0d71a9-0086-494a-91d8-abad0912ba83", "healthMonitor": { "type": "PING", "delay": 10, "timeout": 10, "maxRetries": 10 } } } ]' localhost:7002/openstack/ops
+   */
+  @Override
+  Map operate(List priorOutputs) {
+    task.updateStatus UPSERT_LOADBALANCER_PHASE, "Initializing upsert of load balancer ${description.id ?: description.name} in ${description.region}..."
+    String subnetId = description.subnetId
+    Region region = new Region(name: description.region)
+    LoadBalancerPool newLoadBalancerPool = new LoadBalancerPool(
+      id: description.id,
+      name: description.name,
+      protocol: description.protocol,
+      method: description.method,
+      subnetId: description.subnetId,
+      internalPort: description.internalPort)
+
+    OpenstackClientProvider openstackClientProvider = this.description.credentials.provider
+
+    if (openstackClientProvider.validateSubnetId(region, subnetId)) {
+      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Unable to retrieve referenced subnet ${subnetId} in ${region.name}."
+      throw new OpenstackOperationException(AtomicOperations.UPSERT_LOAD_BALANCER)
+    }
+
+    LbPool resultPool
+    if (newLoadBalancerPool.id) {
+      resultPool = openstackClientProvider.getLoadBalancerPool(region, newLoadBalancerPool.id)
+    }
+
+    if (resultPool) {
+      task.updateStatus UPSERT_LOADBALANCER_PHASE, "Found load balancer pool ${newLoadBalancerPool.id} in ${region.name}."
```

Comment:
Am I reading this right in that it is not updating a load balancer, just creating one?